doc/memos/rdm-draft-alamos-ieee802154-radio-hal.md

```diff
@@ -0,0 +1,967 @@
+     *        frame pending bit, status, etc).
+     *
+     * @pre the device is on
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       frame retransmissions
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[out] info the TX information
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*get_tx_status)(ieee802154_dev_t *dev, ieee802154_tx_info_t *info);
+
+    /**
+     * @brief Set IEEE802.15.4 address in hardware address filter
+     *
+     * @pre the device is on
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       hardware address filtering.
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] short_addr the IEEE802.15.4 short address
+     * @param[in] ext_addr the IEEE802.15.4 extended address
+     * @param[in] pan_id the IEEE802.15.4 PAN ID
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_hw_addr_filter)(ieee802154_dev_t *dev, uint8_t *short_addr,
+                              uint8_t *ext_addr, uint16_t pan_id);
+
+    /**
+     * @brief Set number of frame retransmissions
+     *
+     * @pre the device is on
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       frame retransmissions
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] retries the number of retransmissions
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_frame_retries)(ieee802154_dev_t *dev, int retries);
+
+    /**
+     * @brief Set IEEE802.15.4 CSMA configuration parameters
+     *
+     * @pre the device is on
+     *
+     * @note this function pointer can be NULL if the device doesn't support
+     *       frame retransmissions
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[in] bd parameters of the exponential backoff
+     * @param[in] retries number of CSMA-CA retries. If @p retries < 0, 
+     *                    retransmissions with CSMA-CA MUST be disabled.
+     *
+     * @return 0 on success
+     * @return negative errno on error
+     */
+    int (*set_csma_params)(ieee802154_dev_t *dev, ieee802154_csma_be_t *bd,
+                           int8_t retries);
+```
+
+## 6.2 Event Notification
+
+The Event Notification mechanism is implemented with a function callback.
+The callback function is supposed to be implemented by the upper layer.
+
+The callback signature, the events and their expected behavior are defined
+in the following block:
+
+```c
+/**
+ * @brief   Event notification status
+ */
+typedef enum {
+    /**
+     * @brief the transceiver detected a valid Start Frame Delimiter, which
+              indicates the radio started to receive a packet.
+     */
+    IEEE802154_RADIO_RX_START,
```

Comment:
Could we add some kind of `CRC_ERROR` indication? That way an upper layer can now that a packets that triggered a `RX_START` was discarded, this could also be a `FRAME_DROPPED_INDICATION` if there are other seasons besides CRC to drop the frame.

Reply:
Actually I think this would make the most sense. But in order to introduce that we would need to add the following IMO:
- A CAP to indicate that the radio drops frames with wrong CRC
- If the CAP is there, leave dropping/reading the frame with wrong CRC to the user (this means someone would need to call `indication_rx`).

Rationale:
```
+-----------+-----------+----------+
|   Radio   |    CRC    | AddrFilt |
+-----------+-----------+----------+
| AT86rf2xx | Drop      | Drop     |
| CC2538    | FB (Peek) | Drop     |
| CC2420    | FB (Peek) | Drop     |
| KW2xrf    | Drop      | Drop     |
| mrf24j40  | Drop      | Drop     |
| at86rf215 | Drop      | Drop     |
| nrf52840  | Register  | -        |
+-----------+-----------+----------+
```

Radios that drop frames with wrong CRC won't generate any kind of interrupt. This means, there will be RX_START events without CRC error.
For the radios that have a mechanism to indicate that, the user could still be interested in getting rx info (RSSI, broken MSDU, etc). Also, this offload the "RX flush" on radios that process these events during ISR (`nrf802154`, `cc2538`), etc.

Also an indication is still the best way to go because unifies a bit the code between different radios and also gives information before calling `dev->driver->len` (otherwise a stack could always try to allocate frames with wrong CRC)

Thoughts?

Reply:
> Rationale:
> ```
> +-----------+-----------+----------+
> |   Radio   |    CRC    | AddrFilt |
> +-----------+-----------+----------+
> | AT86rf2xx | Drop      | Drop     |
> | CC2538    | FB (Peek) | Drop     |
> | CC2420    | FB (Peek) | Drop     |
> | KW2xrf    | Drop      | Drop     |
> | mrf24j40  | Drop      | Drop     |
> | at86rf215 | Drop      | Drop     |
> | nrf52840  | Register  | -        |
> +-----------+-----------+----------+
> ```
> 

I believe some of the radios marked as "drop" can actually report frames with invalid CRC in specific configurations, no?

Otherwise having it as an indication is fine with me!

Reply:
Yes, they can be configured. They would fallback to the case of reading a register or reading that information from the framebuffer.

Reply:
Good by me, will you add this to the RDM then?

Reply:
sure!

Reply:
I still have some question regarding this. Some radios will only report invalid CRC in specific modes, in some cases these "modes" would be compile time defined, like for `at86rf2xx` `BASIC_MODE`, but for others it will depend is the radio is for examples in `PROMISCUOS` mode or no. How should we handle this?

Reply:
IMHO `AT86RF2XX_BASIC_MODE` should be removed.
The same can be achieved using the normal driver options.

Reply:
Is something like this what you had in mind @jia200x:

<details><summary>ieee802154/radio.h patch</summary>

```diff
diff --git a/sys/include/net/ieee802154/radio.h b/sys/include/net/ieee802154/radio.h
index f831b84e6f..7d18f15a43 100644
--- a/sys/include/net/ieee802154/radio.h
+++ b/sys/include/net/ieee802154/radio.h
@@ -92,6 +92,10 @@ typedef enum {
      * @brief the device support the IEEE802.15.4 Sub GHz band
      */
     IEEE802154_CAP_SUB_GHZ,
+    /**
+     * @brief the device reports reception off frames with invalid CRC.
+     */
+    IEEE802154_CAP_IRQ_CRC_ERROR,
     /**
      * @brief the device reports when the transmission is done
      */
@@ -186,6 +190,26 @@ typedef enum {
      */
     IEEE802154_RADIO_INDICATION_RX_START,
 
+    /**
+     * @brief the transceiver received a frame with an invalid crc
+     *        and lies in the internal framebuffer.
+     *
+     * This indication should be generated only if the CRC is invalid and
+     * the frame passes the address matching filter (this includes ACK
+     * and Beacon frames). The latter only applies if the radio is not
+     * in promiscuous mode.
+     *
+     * The transceiver is in @ref IEEE802154_TRX_STATE_RX_ON state when
+     * this function is called, but with framebuffer protection (either
+     * dynamic framebuffer protection or disabled RX). Thus, the frame
+     * won't be overwritten before calling @ref ieee802154_radio_indication_crc_error
+     * function. However, @ref ieee802154_radio_indication_crc_error
+     * MUST be called in order to receive new frames. If there's no
+     * interest in the frame, the function can be called with a NULL
+     * buffer to drop the frame.
+     */
+    IEEE802154_RADIO_INDICATION_CRC_ERROR,
+
     /**
      * @brief the transceiver received a frame and lies in the
      *        internal framebuffer.
@@ -233,6 +257,18 @@ typedef struct {
     uint8_t max;    /**< maximum value of the exponential backoff */
 } ieee802154_csma_be_t;
 
+/**
+ * @name IEE802154 rx_info flags
+ *
+ * These flags apply to the received frame
+ * @{
+ */
+/**
+ * @brief   Frame checksum was valid on receive
+ */
+#define IEE802154_RX_INFO_FLAGS_CRC_VALID  (1 << 0)
+/** @} */
+
 /**
  * @brief RX information associated to a frame
  */
@@ -245,6 +281,7 @@ typedef struct {
      */
     uint8_t rssi;
     uint8_t lqi;    /**< LQI of the received frame */
+    uint8_t flags;  /**< flags as defined above */
 } ieee802154_rx_info_t;
 
 /**
@@ -426,6 +463,32 @@ struct ieee802154_radio_ops {
      */
     int (*len)(ieee802154_dev_t *dev);
 
+    /**
+     * @brief Process the CRC Errorindication
+     *
+     * This function reads the received frame from the internal framebuffer.
+     * It should try to copy the received frame into @p buf and
+     * then unlock the framebuffer (in order to be able to receive more
+     * frames).
+     *
+     * @pre the device is on and an @ref IEEE802154_RADIO_INDICATION_CRC_ERROR
+     *      event was issued.
+     *
+     * @post the state is @ref IEEE802154_TRX_STATE_RX_ON
+     *
+     * @param[in] dev IEEE802.15.4 device descriptor
+     * @param[out] buf buffer to write the received PSDU frame into. If NULL,
+     *             the frame is not copied.
+     * @param[in] size size of @p buf
+     * @param[in] info information of the received frame (LQI, RSSI). Can be
+     *            NULL if this information is not needed.
+     *
+     * @return number of bytes written in @p buffer (0 if @p buf == NULL)
+     * @return -ENOBUFS if the frame doesn't fit in @p
+     */
+    int (*indication_crc_error)(ieee802154_dev_t *dev, void *buf, size_t size,
+                                ieee802154_rx_info_t *info);
+
     /**
      * @brief Process the RX done indication
      *
```
</detailes>

Reply:
> IMHO AT86RF2XX_BASIC_MODE should be removed.
> The same can be achieved using the normal driver options.

@jia200x has started re-working the at86rf2xx radio driver I believe, do you agree with this?

Reply:
> Is something like this what you had in mind @jia200x:

Hmmm yes, but I think I would use the same `indication_rx` instead of a dedicated one.
The rationale is that `indication_rx` process the RX_DONE indication, regardless of the CRC status.

The `flags` member seems ok to me. The event definition seems fine, but I think it's important to add that such event only exists if `IEEE802154_RADIO_INDICATION_CRC_ERROR` is there.
Also, I think we would need to define what happens with the address filter there (if the CRC is wrong the address might also be wrong, so most radios only use the filter when the CRC is ok)

> @jia200x has started re-working the at86rf2xx radio driver I believe, do you agree with this?

Definitely! We should remove it.

Reply:
How should this be adapted now that there is a `read` function instead? Does the radio stop listening after receiving a frame with invalid CRC? I would say yes and that the upper layer should reset RX_ON.

Reply:
I adapted OpenWSN to use this, I'll test out the logic and then PR the proposal.

Reply:
I originally wrote there that a failed frame (CRC not OK) would simply continue in `RX_ON`, as most radios already to that. But this could be replaced if it makes the logic on top easier

Reply:
> I originally wrote there that a failed frame (CRC not OK) would simply continue in RX_ON, as most radios already to that. But this could be replaced if it makes the logic on top easier

I was thinking this could be the case if there is no `IEEE802154_CAP_IRQ_CRC_ERROR`, while if it is supported the upper layer would simply have a case:

```c
    switch (status) {
    ...    
    case IEEE802154_RADIO_INDICATION_CRC_ERROR:
        ieee802154_radio_request_set_trx_state(openwsn_radio.dev,
                                               IEEE802154_TRX_STATE_RX_ON);
        while (ieee802154_radio_confirm_set_trx_state(openwsn_radio.dev) == -EAGAIN) {}
        break;
    ...
```

Does that make sense? Or does it feel like unnecessary overhead?

Reply:
I think it would make sense, since these lines are already in the drivers anyway (at least for the `cc2538` and `nrf802154`). Radios that drop the CRC won't indicate anything, so they must continue in RX_ON.